### ISSUE TITLE
GitHub Engagement

### ISSUE THREAD

**Issue by ravi-mosaicml** [mosaicml/composer#664]
Let's make it as easy as possible for customers to engage with Composer on GitHub!

For example:

1. Whenever `trainer.fit()` is invoked, print a message asking users to star our library on GitHub and join our Community Slack.
    If a user made it this far, then they (hopefully) realized some value from Composer, and hopefully they can show some appreciation. Let's also add an env variable that users can set to disable printing of this message, and only print the message once per python session (so it won't show up every time during interactive development).
2. Whenever composer raises an exception, let's add a link that the user can click to open a new GitHub issue. Ideally, we can customize the link to auto-populate the issue with the traceback. This should be doable by hooking into `sys.excepthook` -- see this [StackOverflow post](https://stackoverflow.com/questions/6234405/logging-uncaught-exceptions-in-python) for an example. The user would of course be able to modify the issue before it is submitted to provide any additional context or remove confidential information. (Jupyter will require special handling -- see this [StackOverflow post](https://stackoverflow.com/questions/1261668/cannot-override-sys-excepthook))

CC @kobindra

**Comment by dblalock** [comment]
+1 for (2). I'd add that ideally we should link to search results for that error, rather than the new issue button. My experience in the [Bolt repo](https://github.com/dblalock/bolt) has been that people often open the same issue over and over again.

I'm skeptical of (1) unless we can consistently make it show up only once without the user having to opt out. Personally I would be pretty annoyed if I had to add a line to my `bashrc` to make torch or numpy not nag me about something. (And even adding such a line won't fix this when I ssh into a different server, fire up a new colab notebook, etc).

Or maybe we prompt them to star us when training finishes, because that might be a less frequent and happier moment at which to intercept them.

**Comment by ravi-mosaicml** [comment]
Discussed offline with @bandish-shah; will add that to the File Logger and TQDM Logger rather than to the calls to trainer.fit()

**Comment by bandish-shah** [comment]
Regarding (2) I'm skeptical about making it _too_ easy for users to file issues.  My concern is this will generate a significant amount of noise (especially for issues that we root cause to user errors) that will cause churn.  

What we can do instead is dump the trace to a debug file and provide the user a URL to our GitHub issues page which opens a "User Issue" template.  The user must fill out the form and either copy/paste the output of the debug trace file or attach the file.  This slight overhead is likely enough to incentivize users to not blindly file issues on every error encountered.

Certain errors could trigger automated reporting akin to the "Send diagnostic information" or "Send crash report" functionality commonly seen in OS's.  For example internal `assert`s that are high probability indicators of internal bugs can automatically create an error report and send us relevant information.  Open to dialog on whether or not it makes sense to do that now.  My initial stance is that would still be more beneficial down the road when we have a well defined error handling architecture (error categories, traps, handling) and this is implemented in the library consistently.

**Comment by hanlint** [comment]
I like the intent here, but (1) is too heavy-handed and not appropriate for this library and should not be pursued.

**Comment by hanlint** [comment]
Closing and moving this discussion offline.
